### PR TITLE
refactor: Balance Model and Decay Rounding

### LINKED ISSUE #1608: 🐛 [Bug] Decay in Frontend is Error Prone
## 🐛 Bugreport

image

relates #1596

### PR DESCRIPTION
**Pull request by Mogge**
## 🍰 Pullrequest
Calibrate the decay since the last transaction.

### Issues
- fixes #1608 

### Todo

- [X] None

### CODE REVIEW

File: backend/src/graphql/model/Decay.ts [hunk 1]
```diff
@@ -1,20 +1,24 @@
...
 import Decimal from 'decimal.js-light'
 
+interface DecayInterface {
+  balance: Decimal
+  decay: Decimal
+  roundedDecay: Decimal
+  start: Date | null
+  end: Date | null
+  duration: number | null
+}
```
Comment (ulfgebhardt) [inline backend/src/graphql/model/Decay.ts:11]: Why did you do this? Doesn't this make the code more complex to read with zero benefit? You don't export this interface...

File: backend/src/graphql/model/Decay.ts [hunk 2]
```diff
@@ -1,20 +1,24 @@
...
-    start: Date | null,
-    end: Date | null,
-    duration: number | null,
-  ) {
-    this.balance = balance
-    this.decay = decay
-    this.start = start
-    this.end = end
-    this.duration = duration
+  constructor(data: DecayInterface) {
```
Comment (ulfgebhardt) [inline backend/src/graphql/model/Decay.ts:15]: you can make `{ balance: Deciaml, ... }`  curly brackets around the parameters to archive the same thing.

File: backend/src/graphql/model/Transaction.ts [hunk 3]
```diff
@@ -17,21 +17,26 @@ export class Transaction {
...
-        transaction.decay.toDecimalPlaces(2, Decimal.ROUND_FLOOR),
-        transaction.decayStart,
-        transaction.balanceDate,
-        Math.round((transaction.balanceDate.getTime() - transaction.decayStart.getTime()) / 1000),
-      )
+      this.decay = new Decay({
+        balance: transaction.balance.toDecimalPlaces(2, Decimal.ROUND_DOWN),
+        decay: transaction.decay.toDecimalPlaces(2, Decimal.ROUND_FLOOR),
+        // TODO
+        roundedDecay: new Decimal(0),
```
Comment (ulfgebhardt) [inline backend/src/graphql/model/Transaction.ts:33]: this seems incorrect

File: backend/src/util/virtualTransactions.ts [hunk 4]
```diff
@@ -68,12 +68,12 @@ const virtualDecayTransaction = (
     userId: -1,
     previous: -1,
     typeId: TransactionTypeId.DECAY,
-    amount: decay.decay ? decay.decay.toDecimalPlaces(2, Decimal.ROUND_FLOOR) : new Decimal(0), // new Decimal(0), // this kinda is wrong, but helps with the frontend query
+    amount: decay.decay ? decay.roundedDecay : new Decimal(0), // this kinda is wrong, but helps with the frontend query
```
Comment (ulfgebhardt) [inline backend/src/util/virtualTransactions.ts:71]: can we move this comment to the actual calculation in `util/decay.ts` explaingin why and how we do that, please?

File: frontend/src/layouts/DashboardLayout_gdd.vue [hunk 5]
```diff
@@ -53,11 +53,9 @@ export default {
   },
   data() {
     return {
-      logo: 'img/brand/green.png',
```
Comment (ulfgebhardt) [inline frontend/src/layouts/DashboardLayout_gdd.vue:56]: ????

Comment (Mogge) [reply frontend/src/layouts/DashboardLayout_gdd.vue:56]: Never used

File: backend/src/graphql/model/Transaction.ts (hunk 3 above)
Comment (Mogge) [reply backend/src/graphql/model/Transaction.ts:33]: Therefore to do

File: backend/src/util/virtualTransactions.ts (hunk 4 above)
Comment (Mogge) [reply backend/src/util/virtualTransactions.ts:71]: I do not know what this comment is for. This is exactly the point why I do not like comments in the code. They pollute the code and create confusion.

**Review by ulfgebhardt — APPROVED** [review]

**Pull request merged**